### DISCUSSION TITLE
Support multiple languages

### DISCUSSION THREAD

**Discussion by AndresMorelos** [AndresMorelos/InvoncifyReleases#31] [General]
Hey there, If you want this app to support your language just write it below.

> Please search first if your language is in the comment, if yes, vote up to give it a priority, also, consider to contribute in the translation.

![image](https://user-images.githubusercontent.com/6563162/159202907-3ffbd73c-b123-4f1d-a3c3-dcdff98c75a9.png)

**Comment by sonalder-darlene** [comment]
French (incomplete as for now)
I could contribute to it. How do you handle localization ? Is there a file where I can simply translate english strings into french ?

**Reply by AndresMorelos** [reply]
@sonalder-darlene you can help to translate the app  using the Crowdin project https://crowdin.com/project/invoncify

**Comment by ghost** [comment]
I have translated to Serbian Cyrillic on ~~Crowdwin~~ Crowdin, it should be production ready since I consulted a friend about some local tax code terminology. You can do transliteration to generate 100% accurate Serbian Latin translation due to the nature of the language.

```
SerbianCyr = ['љ', 'њ', 'џ', 'Љ', 'Њ', 'Џ', 'а', 'б', 'в', 'г', 'д', 'ђ', 'е', 'ж', 'з', 'и', 'ј', 'к', 'л', 'м', 'н', 'о', 'п', 'р', 'с', 'т', 'ћ', 'у', 'ф', 'х', 'ц', 'ч', 'ш', 'А', 'Б', 'В', 'Г', 'Д', 'Ђ', 'Е', 'Ж', 'З', 'И', 'Ј', 'К', 'Л', 'М', 'Н', 'О', 'П', 'Р', 'С', 'Т', 'Ћ', 'У', 'Ф', 'Х', 'Ц', 'Ч', 'Ш'];

SerbianLat = ['lj', 'nj', 'dž', 'Lj', 'Nj', 'Dž', 'a', 'b', 'v', 'g', 'd', 'đ', 'e', 'ž', 'z', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'r', 's', 't', 'ć', 'u', 'f', 'h', 'c', 'č', 'š', 'A', 'B', 'V', 'G', 'D', 'Đ', 'E', 'Ž', 'Z', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'R', 'S', 'T', 'Ć', 'U', 'F', 'H', 'C', 'Ć', 'Š'];
```

Basically have an array of Cyrillic characters and substitute for Latin ones. Array elements should be in that order, because some Cyrillic letters are essentially two Latin ones smooshed together. It's not necessary for Cyrillic to Latin, but it is other way around as you want "lj" to be transliterated to љ, not individual л and ј.

When you have the new build I'll go over it again to see the translation in context and correct for some errors if required.

Thank you for your hard work and contributions.

**Reply by AndresMorelos** [reply]
@milejancic this language will be deployed in production at the 1.9.0 version keep posted on PR #52

**Reply by AndresMorelos** [reply]
@milejancic released at https://github.com/AndresMorelos/Invoncify/releases/tag/v1.9.0

**Comment by bdlanduyt** [comment]
I'll translate dutch due crowdin.com.

**Reply by bdlanduyt** [reply]
Translated all. How do I use it?

**Reply by AndresMorelos** [reply]
@bdlanduyt let me review it and create a new release with the language, thank you for your help

**Reply by AndresMorelos** [reply]
@bdlanduyt Dutch Language release at [v1.11.0](https://github.com/AndresMorelos/Invoncify/releases/tag/v1.11.0), thank you again for your help

**Comment by ghost** [comment]
I added missing strings for Serbian Cyrillic and fixed some typos.

**Reply by AndresMorelos** [reply]
@milejancic Ok, let me review it, thank you for your help

**Reply by AndresMorelos** [reply]
@milejancic Serbian Language updated at [v1.11.0](https://github.com/AndresMorelos/Invoncify/releases/tag/v1.11.0), thank you again for your help

**Comment by bdlanduyt** [comment]
How do I get installed?

Op do 14 apr. 2022 18:13 schreef Andrés Morelos ***@***.***>:

**Reply by AndresMorelos** [reply]
@bdlanduyt if you have already installed a previous version of the app just press `command or control + U`, this will check for updates and install the latest one.

**Comment by bdlanduyt** [comment]
I get an error: 'An error has occurred' New version 1.12.0 is nog signed by
the application owner: publisherNames:...
Kind regards

Op vr 15 apr. 2022 om 04:51 schreef Andrés Morelos ***@***.***

**Comment by bdlanduyt** [comment]
I've downloaded the original version from softpedia.com. Maybe that's the
reason? Where can I find a 'real' download?
Kind regards

Op vr 15 apr. 2022 om 04:51 schreef Andrés Morelos ***@***.***

**Reply by AndresMorelos** [reply]
@bdlanduyt from now you can update to the latest version here https://github.com/AndresMorelos/Invoncify/releases/tag/v1.12.0

I'm verifying the issue with the publisherNames

**Comment by bdlanduyt** [comment]
The invoice itself and the previewer are still in English.
It also would be nice if the price of an item could be lower than 0
(ex.deposit).
Kind regards

Op za 16 apr. 2022 om 19:52 schreef Andrés Morelos ***@***.***

**Reply by AndresMorelos** [reply]
@bdlanduyt conversation moved to #60

**Comment by bdlanduyt** [comment]
Is it possible to translate the previewer? Van't pick 'Nederlands'.

Op za 16 apr. 2022 19:52 schreef Andrés Morelos ***@***.***>:

**Reply by AndresMorelos** [reply]
@bdlanduyt could you please download and install the latest version https://github.com/AndresMorelos/Invoncify/releases/tag/v1.13.2

now you're will be able to translate the previewer.

**Comment by ghost** [comment]
Updated Serbian Cyrillic for the next build. No rush tho, nothing major.

**Comment by ghost** [comment]
@AndresMorelos I can do Serbian Latin too if you add it to Crowdin.

**Reply by AndresMorelos** [reply]
@milejancic Added, thank you for your support

**Reply by ghost** [reply]
OK, so my plan was to use gettext utility to [recode Cyrillic to Latin ](https://www.gnu.org/software/gettext/manual/gettext.html#index-recode_002dsr_002dlatin-program) and of course I messed up. Twice. Should have tried on a smaller file first.

What I found out however is:
1. Crowdin is poorly documented, but to be fair maybe I should have gone somewhere else to learn more about .xliff format
2. Crowdin will show me comments on every single language page, not just the one in question. So for example I would have comments like "Is this Serbian translation correct or this one" on a French or Chinese pages. Weird.
3. Crowdin will ignore translations if they are spelled exactly the same. For example "Logo" is spelled the same in Serbian Latin and in English, so it just ignores it as not translated. But it works for Cyrillic, same spelling but different script. 

Both translations should be OK for the next build.

**Reply by ghost** [reply]
Oh, and when you get the chance if you could add Serbian (Cyrillic) and Serbian(Latin) as supported in readme that would be great. We have lots of nationalities who were born in neighboring countries and have never learned Cyrillic script, so knowing there is 100% Latin translation available is great for them.

**Reply by AndresMorelos** [reply]
@milejancic Languages added to the README, Cyrillic updated, and Latin added at v1.30.1

Thank you for your help
